feat: Receive and display reactions on a conversation - WPB-626

agisilaos opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-626" title="WPB-626" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-626</a>  receive and display reactions in conversation view
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we introduce a new way to show reactions under a message. This code adds the ability to show the reactions and interact with the,. A few notes to keep in mind:

- We need to get rid of `MessageToolBoxView` and move things such as the date and time of the message as well as the status of it to the upper right side of the message. This will happen in a separate PR since as you can see this is already a big one. 

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

agisilaos committed:
chore: WIP

agisilaos committed:
chore: push

agisilaos committed:
chore: make the height of the collection view adaptable

agisilaos committed:
chore: groom

agisilaos committed:
chore: Re-organize reaction code and add colors

agisilaos committed:
chore: Organize reactions code and create basic reaction view

agisilaos committed:
chore: commit

agisilaos committed:
chore: cleanup

agisilaos committed:
fix: add action to the reaction view

agisilaos committed:
Merge branch 'feat/reactions' into receive-and-display-reactions-in-conversation-view

agisilaos committed:
chore: Replace ReactionType with MessageReaction

agisilaos committed:
Merge branch 'feat/reactions' into receive-and-display-reactions-in-conversation-view

agisilaos committed:
Merge branch 'feat/reactions' into receive-and-display-reactions-in-conversation-view

agisilaos committed:
fix: Mistake enabling staging and commiting it

agisilaos committed:
fix: Position of the reactionView

It was more on the left side of the screen and with that change we make it more to the right and just below the message. As it is on the design specs

agisilaos committed:
fix: Set the bg color of the collectionView .clear

So we follow the same pattern as with the other views on this format. Also that change fixes a problem we were having with snapshot tests

agisilaos committed:
fix: Update CompositeMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update ConversationAudioMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update ConversationFileMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update TextMessageMentionsTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update ConversationVideoMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update ConversationTextMessageTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update ConversationImageMessageTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
Revert "fix: Update CompositeMessageCellTests"

This reverts commit d9b6c16bbd4d4d085e0a178936bd2baf3c32abfc.

agisilaos committed:
fix: Update CompositeMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update LocationMessageCellTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update MarkDownSnapshotTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: Update MultiParagraphSnapshotTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
fix: SwiftLint Warnings

agisilaos committed:
fix: Update ConversationPingMessageTests

Since we're introducing a new cell (reactions) the message views are going to change a bit regarding layout. With that said we need to update snapshot tests.

agisilaos committed:
chore: Update values of accessibility identifiers

So they can match better the type of the cell

agisilaos committed:
chore: Get rid of unused code

That code was there because of the prototype phase of the feature. We don't need it anymore, therefore we get rid of it.

agisilaos committed:
doc: Add docs to ReactionsCellView

Some documentation to the ReactionsCellView so we can navigate the file faster and with more ease

agisilaos committed:
doc: Add doc to ReactionCollectionViewCell

Some documentation to the ReactionCollectionViewCell so we can navigate the file faster and with more ease

agisilaos committed:
doc: Add doc to ReactionCollectionView

Some documentation to the ReactionCollectionView so we can navigate the file faster and with more ease

agisilaos committed:
doc: Add doc to ReactionToggleButton

agisilaos committed:
doc: Add doc to ReactionCollectionViewCell

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ConversationReactionsCellDescription.swift:
```diff
@@ -0,0 +1,57 @@
+
+    var accessibilityIdentifier: String? = "reaction message cell"
+
+    var accessibilityLabel: String? = "reaction message cell"
```
Voice over reads `accessibilityLabel`, it shouldn't be hardcoded

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,117 @@
+    // MARK: - configure method
+
+    func configure(
+        with object: Configuration,
```
I don't see that you use `object` parameter

agisilaos replied:
I think we already talked about that in person so I'm marking it as resolved.

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,117 @@
+        reactionView.translatesAutoresizingMaskIntoConstraints = false
+        self.translatesAutoresizingMaskIntoConstraints = false
+
+        reactionView.trailingAnchor.constraint(equalTo: self.trailingAnchor).isActive = true
```
Why don't you use `NSLayoutConstraint.activate([]) `?

agisilaos replied:
Done!

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,117 @@
+
+    func configure(
+        with object: Configuration,
+        animated: Bool
```
You don't use `animated` too

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,106 @@
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        collectionView.fitIn(view: self)
+
+        contentSizeObservation = collectionView.observe(\.contentSize, options: .new, changeHandler: { [weak self] (tv, _) in
```
Why do we need this observation? And can you find more readable name for `tv` property?

agisilaos replied:
We need that observation to make sure the height of the cell adapts based on the reactions we show (e.g two lines of reactions)

agisilaos replied:
> And can you find more readable name for tv property?

Done!

KaterinaWire commented:
@agisilaos All new snapshots are irrelevant to your changes, right? Maybe this is due to the replacement of `XCTestCase` with `ZMSnapshotTestCase`?

agisilaos commented:
> @agisilaos All new snapshots are irrelevant to your changes, right? Maybe this is due to the replacement of `XCTestCase` with `ZMSnapshotTestCase`?

@KaterinaWire This is due to adding a new ReactionCellDescription so all Message related tests needed to get updated.

agisilaos commented:
> @agisilaos Do you have tests for your changes?

This is going to happen on a separate PR after we get rid of the `MessageToolBoxView`.

KaterinaWire commented:
> > @agisilaos Do you have tests for your changes?
> 
> This is going to happen on a separate PR after we get rid of the `MessageToolBoxView`.

I use this view for offline backend error messages. We should sync with you, me and Wolfgang before removing this view.

agisilaos committed:
fix: configure message for reactions

With this change we make sure that reactions are shown when we move from conversation to conversation list

aka We configure the cell

agisilaos committed:
fix: SwiftLint warnings

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ConversationReactionsCellDescription.swift:
```diff
@@ -0,0 +1,56 @@
+
+// MARK: - ReactionMessageCellDescription
+
+final class ReactionMessageCellDescription: ConversationMessageCellDescription {
```
Name suggestion: `MessageReactionsCellDescription`

agisilaos replied:
Done!

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ConversationReactionsCellDescription.swift:
```diff
@@ -0,0 +1,56 @@
+
+    weak var actionController: ConversationMessageActionController?
+
+    var accessibilityIdentifier: String? = "reaction message cell"
```
I don't think identifiers should have spaces.

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+
+// MARK: - ReactionsCellView
+
+final class ReactionsCellView: UIView, ConversationMessageCell {
```
Suggestion: `MessageReactionsCell`

agisilaos replied:
Done!

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+
+// MARK: - Reaction
+
+public struct Reaction {
```
Suggestion: `MessageReactionMetadata`

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+public struct Reaction {
+
+    let type: MessageReaction
+    let count: Int
```
Suggestion: `UInt`

agisilaos replied:
Why though? Since we know the the count it's going to be an integer!?

johnxnguyen replied:
It make sure that we don't pass in a negative number by accident.

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+    // MARK: - Properties
+
+    struct Configuration: Equatable {
+        let message: ZMConversationMessage
```
You should only use what you need, which is the reaction metadata.

agisilaos replied:
Done!

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    let reactionView = ReactionCollectionView()
```
Suggestion: `reactionCollectionView`

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+
+    let reactionView = ReactionCollectionView()
+
+    var arrayReactions: [Reaction] = []
```
Suggestion: `reactions`

agisilaos replied:
It's not used anywhere, so I'm getting rid of it.

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCellView.swift:
```diff
@@ -0,0 +1,119 @@
+        animated: Bool
+    ) {
+
+        reactionView.reactions = object.message.usersReaction.compactMap { reaction, usersWhoReacted in
+            guard
+                let reactionType = MessageReaction.messageReaction(from: reaction),
+                !usersWhoReacted.isEmpty
+            else {
+                return nil
+            }
+
+            return Reaction(
+                type: reactionType,
+                count: usersWhoReacted.count,
+                isSelfUserReacting: usersWhoReacted.contains(where: \.isSelfUser),
+                performReaction: { [weak self] in
+                    guard let `self` = self else { return }
+                    self.delegate?.perform(
+                        action: .react(reactionType),
+                        for: object.message,
+                        view: self
+                    )
+                }
+            )
+        }
```
This seems like a lot of logic for a view. I suggest to try move this code elsewhere, and simply inject it in.

agisilaos replied:
Done!

agisilaos committed:
chore: Use NSLayoutConstraint.activate

Instead of activating each constraint individually

agisilaos committed:
chore: Replace tv with cell

For better readability of the property

agisilaos commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/ConversationMessageSectionController.swift:
```diff
@@ -242,6 +242,10 @@ final class ConversationMessageSectionController: NSObject, ZMMessageObserver {
         }
 
+        if !message.isSystem {
+            add(description: ReactionMessageCellDescription(message: message))
```
That's a valid question. I can do some testing and see how it will show.

agisilaos replied:
I've updated the code and the changes in snapshots aren't needed anymore. Thank you for pointing that out @KaterinaWire!

agisilaos committed:
chore: Update Snapshot Tests

Since we're planning to show the reactions cell only when reactions exist on the message we didn't need the changes on the snapshots after all.

agisilaos committed:
fix: Update LocationMessageCellTests

Since we're planning to show the reactions cell only when reactions exist on the message we didn't need the changes on the snapshots after all.

agisilaos committed:
chore: Improve the naming of properties

agisilaos committed:
chore: Rename of class

from ReactionMessageCellDescription to MessageReactionsCellDescription

agisilaos committed:
chore: Rename of class

From ReactionsCellView to MessageReactionsCell

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,106 @@
+
+    // MARK: - Properties
+
+    lazy var collectionViewHeightConstraint: NSLayoutConstraint = collectionView.heightAnchor.constraint(equalToConstant: 40)
```
Where does 40 come from?

agisilaos replied:
From Figma[https://www.figma.com/file/IBRRhYC0Xq10vy33mewhGG/Reactions%2FEmojis?type=design&node-id=1651%3A5583&t=K9iSdS5HVg9hhTI8-1]

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: 51, height: 24)
```
Where do these numbers come from?

agisilaos replied:
From Figma[https://www.figma.com/file/IBRRhYC0Xq10vy33mewhGG/Reactions%2FEmojis?type=design&node-id=1651%3A5583&t=K9iSdS5HVg9hhTI8-1]

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionToggleButton.swift:
```diff
@@ -0,0 +1,119 @@
+
+// MARK: - ReactionToggleButton
+
+class ReactionToggleButton: UIControl {
```
Suggestion: `ReactionToggle`

agisilaos replied:
Done!

agisilaos committed:
fix: Add some SnapshotTests for Reactions

agisilaos committed:
chore: Remove recordMode = true

agisilaos committed:
fix: Update test for multiple reactions

agisilaos committed:
chore: remove recordMode

Since it's not needed

agisilaos committed:
fix: Rename ReactionToggleButton

to ReactionToggle

agisilaos commented:
> @agisilaos Do you have tests for your changes?

I've updated this PR with Tests!

agisilaos committed:
chore: Groom and cleanup

Made the code much simple. It resulted needing less dependencies on the tests among other things.

Co-Authored-By: John Nguyen <28632506+johnxnguyen@users.noreply.github.com>

agisilaos committed:
fix: Update Tests

agisilaos commented:
@KaterinaWire & @johnxnguyen The PR is ready for review..

johnxnguyen commented:
> This is going to happen on a separate PR after we get rid of the `MessageToolBoxView`.

@agisilaos why do we need to do tests in a separate PR? Generally I think it's good to try to add the tests together with the code being tested in a single PR.

johnxnguyen reviewed (commented):
Did you test voice over with this new UI? How does it behave?

johnxnguyen commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCellDescription.swift:
```diff
@@ -0,0 +1,74 @@
+
+    weak var actionController: ConversationMessageActionController?
+
+    var accessibilityIdentifier: String? = "reaction message cell"
```
What about the identifier with spaces?

agisilaos replied:
Done!

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCell.swift:
```diff
@@ -0,0 +1,114 @@
+
+        NSLayoutConstraint.activate([
+            reactionCollectionView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
+            reactionCollectionView.leadingAnchor.constraint(equalTo: self.leadingAnchor, constant: 40),
```
Do we have a constant for this offset? The leading constraint should match the main message's constraint. If we have one, it would be nice to use the same value instead of defining it in a very similar view.

KaterinaWire replied:
Or maybe it's better to have this constraint for the MessageReactionsCell view

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCell.swift:
```diff
@@ -0,0 +1,114 @@
+        animated: Bool
+    ) {
+        reactionCollectionView.reactions = object.reactions.map {
+            var reaction = $0
```
Can't you do  just `object.reactions.map { reaction in` ?

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCell.swift:
```diff
@@ -0,0 +1,114 @@
+                    return
+                }
+
+                self.delegate?.perform(
```
@johnxnguyen @agisilaos  Why are we doing this in the view? Shouldn't view be a simple? Why not configure the perform action in the Description?

johnxnguyen replied:
We do, but here we're adding the action which needs to use the delegate. As far as I know, the delegate is only available in the view.

KaterinaWire replied:
`MessageReactionsCellDescription` has `delegate: ConversationMessageCellDelegate` too (but I'm not sure is it nil or not at that moment). We create reactions in `MessageReactionsCellDescription` and send them to the `MessageReactionCell` with nil actions. And configure actions inside MessageReactionCell.

agisilaos replied:
That's not possible because we don't have a view on the MessageReactionsCellDescription and the view parameter is not optional on the perform method. With that said I'm marking that discussion as resolved.

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCellDescription.swift:
```diff
@@ -0,0 +1,74 @@
+
+    var accessibilityIdentifier: String? = "reaction message cell"
+
+    var accessibilityLabel: String? = "reaction message cell"
```
Do you have a ticket for voice over?

agisilaos replied:
> Do you have a ticket for voice over?

No but I can create one!

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,108 @@
+
+    // MARK: - Properties
+
+    lazy var collectionViewHeightConstraint: NSLayoutConstraint = collectionView.heightAnchor.constraint(equalToConstant: 40)
```
I left a comment about this magic 40 for another view. Looks like we use it in several places. Why don't create a property to it?

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionToggle.swift:
```diff
@@ -0,0 +1,119 @@
+        stackView.spacing = 4
+
+        stackView.isLayoutMarginsRelativeArrangement = true
+        stackView.layoutMargins = UIEdgeInsets(top: 4, left: 8, bottom: 4, right: 0)
```
Why do you need this margin?

agisilaos replied:
For adding the proper spacing within the stack view where the emoji and counter lives.

KaterinaWire replied:
Why not to add this to main constrains in `NSLayoutConstraint.activate([ ...])`? I think it will be clearer.

KaterinaWire replied:
Why is the right insert 0 ?

agisilaos replied:
That's has been fixed!

agisilaos committed:
fix: Spacing of reactions

agisilaos commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: 51, height: 24)
```
Fixed!

KaterinaWire commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCellDescription.swift:
```diff
@@ -0,0 +1,74 @@
+                type: reactionType,
+                count: UInt(usersWhoReacted.count),
+                isSelfUserReacting: usersWhoReacted.contains(where: \.isSelfUser),
+                performReaction: nil
```
You have a delegate here, line #67

agisilaos committed:
chore: Remove spaces from accessibilityIdentifier

agisilaos committed:
Merge branch 'feat/reactions' into receive-and-display-reactions-in-conversation-view

agisilaos committed:
chore: Create a property for the height value

Avoid using magic number and create properties instead

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,113 @@
+        return UICollectionView(frame: .zero, collectionViewLayout: self.flowLayout)
+    }()
+
+    var reactions = [MessageReactionMetadata]() {
```
Can this be private ? How do we make sure it's called on mainThread, see didSet reloadData()

agisilaos replied:
No, it can't be private since it's being used on MessageReactionsCell.

```swift
func configure(
        with object: Configuration,
        animated: Bool
    ) {
        reactionCollectionView.reactions = object.reactions.map {
            var reaction = $0

            reaction.performReaction = { [weak self] in
                guard
                    let `self` = self,
                    let message = self.message
                else {
                    return
                }

                self.delegate?.perform(
                    action: .react(reaction.type),
                    for: message,
                    view: self
                )
            }

            return reaction
        }

    }
```

netbe replied:
just make sure then that it's always called on mainThread or add a DispatchQueue.main.async in the setter.

agisilaos replied:
Done!

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/MessageReactionsCell.swift:
```diff
@@ -0,0 +1,114 @@
+
+    // MARK: - configure method
+
+    func configure(
```
apart from tests where is this called ?

agisilaos replied:
This is part of the `ConversationMessageCell` protocol.

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,113 @@
+    // MARK: - Properties
+
+    private var collectionViewHeight: CGFloat = 40
+    lazy var collectionViewHeightConstraint: NSLayoutConstraint = collectionView.heightAnchor.constraint(equalToConstant: collectionViewHeight)
```
why lazy if you assign it now?

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,113 @@
+
+    private var collectionViewHeight: CGFloat = 40
+    lazy var collectionViewHeightConstraint: NSLayoutConstraint = collectionView.heightAnchor.constraint(equalToConstant: collectionViewHeight)
+    let flowLayout = UICollectionViewFlowLayout()
```
private ?

agisilaos replied:
Done!

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionView.swift:
```diff
@@ -0,0 +1,113 @@
+    // MARK: - Properties
+
+    private var collectionViewHeight: CGFloat = 40
+    lazy var collectionViewHeightConstraint: NSLayoutConstraint = collectionView.heightAnchor.constraint(equalToConstant: collectionViewHeight)
```
private ?

agisilaos replied:
Done!

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionCollectionViewCell.swift:
```diff
@@ -0,0 +1,66 @@
+        )
+    }
+
+    func addViews() {
```
private

agisilaos replied:
Done!

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionToggle.swift:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private func updateAppearance() {
+        backgroundColor = isToggled
+        ? ButtonColors.backgroundReactionSelected :
+        ButtonColors.backroundReactionNormal
```
Don't go to next line in middle of ternary
```suggestion
        backgroundColor = isToggled ? ButtonColors.backgroundReactionSelected : ButtonColors.backroundReactionNormal
```

netbe commented on wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Components/Reactions/ReactionToggle.swift:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    private func updateAppearance() {
+        backgroundColor = isToggled
+        ? ButtonColors.backgroundReactionSelected :
+        ButtonColors.backroundReactionNormal
+
+        layer.borderColor = isToggled
+        ? ButtonColors.borderReactionSelected.cgColor :
+        ButtonColors.borderReactionNormal.cgColor
+
+        counterLabel.textColor = isToggled
+        ? SemanticColors.Label.textReactionCounterSelected :
+        SemanticColors.Label.textDefault
```
or better:
```suggestion
        if isToggled {
           backgroundColor = ButtonColors.backgroundReactionSelected
           layer.borderColor = ButtonColors.borderReactionSelected.cgColor
           counterLabel.textColor = SemanticColors.Label.textReactionCounterSelected
        } else {
           backgroundColor = ButtonColors.backroundReactionNormal
           layer.borderColor = ButtonColors.borderReactionNormal.cgColor
           counterLabel.textColor = SemanticColors.Label.textDefault
        }        
```

agisilaos committed:
chore: Improve readability of code

Co-authored-by: François Benaiteau <netbe@users.noreply.github.com>

agisilaos commented on wire-ios/Wire-iOS Tests/ReferenceImages_64/Wire_iOS_Tests.ConversationReactionMessageTests/testThatItConfiguresWithOtherReactions@2x.png:
I know, I'm trying to find a way to fix that.

agisilaos replied:
Fixed with my latest commit!

agisilaos committed:
adress code review

agisilaos committed:
fix: SwiftLint warnings

agisilaos committed:
fix: SwiftLint warnings

agisilaos committed:
fix: Use a property instead of a magic number

agisilaos committed:
chore: Call it on the main thread

agisilaos committed:
fix: error

agisilaos committed:
fix: layout issue with reactions
